Mark threads for suspend and continue exitting.

ivyl opened the pull request:
If one of the threads is in a native call it won't be able to cooperatively suspend itself.

This may happen e.g. for services using WinSW wrapper which end up StartServiceCtrlDispatcherW() via a call to
System.ServiceProcess.ServiceBase.Run(). The wrapper may tray to call System.Environment.Exit() if the wrapped process exits with a non-zero exit code.

madewokherd commented:
I'm not sure whether it's safe to move on from this point. If it is safe, then we should just move on after 1 attempt.

The point of co-op suspend was supposed to be that we can just ignore threads that aren't in managed code. It should be possible to just set some state that tells the thread to suspend if it ever makes its way to a managed code entry point, and then move on. Similarly, they should be able to prevent any new managed threads from being created, or existing unmanaged threads from running managed code. This whole retry loop seems ill-advised to me.

ivyl commented:
I've read more of the shutdown code, as well as `docs/threading` and followed the history of changes to `mono_thread_suspend_all_other_threads()`. Looks like it was reworked heavily but never completely cleaned up.

The biggest thing was addition of the `shutting_down` "barrier". This guarantees no new threads will be started so we don't need to keep on looping over threads.

I've also fixed up the version I've force-pushed yesterday. I've corrected some typos and removed duplicated early-return conditions in `suspend_all_other_threads()`

madewokherd commented:
Would you expect this to work OK on non-Windows platforms? I'm considering committing it to the main branch on Winehq's fork, then merging it into wine-mono.

ivyl commented:
Yes. I see no reason why it would become problematic on other platforms.

madewokherd commented:
I tried rebasing this onto https://gitlab.winehq.org/wine-mono/mono/-/tree/main?ref_type=heads and testing a Linux build, and I'm getting some test failures when running `make -C mono/tests test-jit` that I think may be new.

```
Failed test(s):

finalizer-exit.exe
=============== finalizer-exit.exe.stdout ===============
Set ID: foo
finalizer thread ID: foo
ghashtable.c:334: assertion 'hash != NULL' failed

* Assertion at threads.c:637, condition `offset' not met


=============== EOF ===============
=============== finalizer-exit.exe.stderr ===============

=============== EOF ===============
```

madewokherd commented on mono/metadata/threads.c:
```diff
@@ -3921,84 +3938,31 @@ void mono_thread_suspend_all_other_threads (void)
 		mono_threads_unlock ();
+		mono_thread_info_sleep (50, NULL);
+		timeout -= mono_msec_ticks() - start_time;
+	}
```
I'm not sure, but I think this counts the time quadratically? You're subtracting time since the start each iteration. I think it'd make more sense to store the end time and compare to that.

ivyl replied:
Correct, I forgot to stage a line that was reassigning `start_time = mono_msec_ticks ();`. I'll push that in a moment.

madewokherd replied:
It's a bit nit-picky, but I feel like this wait loop is way more complicated than it needs to be. I'm hung up on there being two calls to `mono_msec_ticks` in the loop, which shouldn't make a practical difference but in theory there could be time passing between them. You could just assign start_time once and check `mono_msec_ticks() - start_time < 1000`.

madewokherd replied:
Also, if I'm reading this right, we always wait 50 ms, even if `still_suspending` was set to FALSE on this iteration. The final iteration will wait, but it won't check whether the threads finished suspending.

I think I would add a helper function for checking for suspended threads and write the loop something like this:

```c
start_time = mono_msec_ticks ();
while ((still_suspending = is_still_suspending ()) && mono_msec_ticks () - start_time < 1000)
{
    mono_thread_info_sleep (50, NULL);
}

if (still_suspending) emit warning and skip shutdown
```

ivyl replied:
I've based the original logic on some other timeouts implemented in the same file. I've changed the implementation to match your suggestion.

madewokherd commented:
If it's really impossible to clean up safely, then I think we should skip cleanup in that case.

madewokherd commented on mono/metadata/icall.c:
```diff
@@ -7299,13 +7299,27 @@ ves_icall_System_Environment_Exit (int result)
+
+#ifdef HOST_WIN32
+	/*
+	 * Wine msvcrt's exit() calls CorExitProcess() if mscoree is loaded.
```
Wine's `CorExitProcess` is supposed to detect that mono has begun shutting down and call `ExitProcess` in that case, but it seems detecting that relies on `mono_profiler_set_runtime_shutdown_begin_callback`, which I'm guessing is part of `mono_runtime_quit_internal`.

Looking at it again now, it probably would've been better to check `mono_runtime_is_shutting_down`, but since we need to maintain compatibility with current Wine, I wouldn't want Wine Mono (at least for 9.x) to depend on that. So I think we'll want to keep this for now as a Wine Mono specific hack.

ivyl replied:
I've split this into a separate commit so you don't have to apply it over to Mono repo and keep it local to Wine-Mono.

ivyl committed:
Mark threads for suspend and continue exitting soon after.

New threads cannot be created due to the `shutting_down` barrier and any
existing thread will get suspended as soon as they try to continue their
managed execution. We have to wait for them to reach suspended state as
otherwise they will race against cleanup done by
mono_runtime_quit_internal() and may reference already tore down data
structure.

Sadly, if one of the threads is in a native call it won't be able to
enter suspend at all.

This may happen e.g. for services using WinSW wrapper which end up
StartServiceCtrlDispatcherW() via a call to
System.ServiceProcess.ServiceBase.Run(). The wrapper may tray to call
System.Environment.Exit() if the wrapped process exits with a non-zero
exit code.

We should wait some sensible amount of time before giving up and trying
continuing the shutdown anyway.

A lot of the old code was there from an older implementation that was
calling `wait_for_tids()`. It's been reworked multiple times but  was
never fully cleaned up. Turns there's no need to open the handles and
have to care about the MONO_W32HANDLE_MAXIMUM_WAIT_OBJECTS limit any
more.

ivyl committed:
WINE-MONO HACK: Use ProcessExit() in System.Environment.Exit

madewokherd commented:
Thanks, I'll merge once I confirm the test results for this revision.

madewokherd commented:
Pushed midstream as https://gitlab.winehq.org/wine-mono/mono/-/commit/05dbc5cc24d96e7739e883be41780f73847bd8cb

madewokherd commented:
And the hack in Wine Mono: https://github.com/madewokherd/mono/commit/3aeb8c115c55628762642b9d1feffb39450798f0
